piq/ssim.py

```diff
@@ -150,213 +161,28 @@ def __init__(self, kernel_size: int = 11, kernel_sigma: float = 1.5, k1: float =
-           Multi-scale Structural Similarity for Image Quality Assessment.
-           IEEE Asilomar Conference on Signals, Systems and Computers, 37,
-           https://ieeexplore.ieee.org/document/1292216
-           :DOI:`10.1109/ACSSC.2003.1292216`
-        .. [2] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P.
-           (2004). Image quality assessment: From error visibility to
-           structural similarity. IEEE Transactions on Image Processing,
-           13, 600-612.
-           https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf,
-           :DOI:`10.1109/TIP.2003.819861`
-    """
-    __constants__ = ['kernel_size', 'k1', 'k2', 'sigma', 'kernel', 'reduction']
-
-    def __init__(self, kernel_size: int = 11, kernel_sigma: float = 1.5, k1: float = 0.01, k2: float = 0.03,
-                 scale_weights: Optional[Union[Tuple[float], List[float], torch.Tensor]] = None,
-                 reduction: str = 'mean', data_range: Union[int, float] = 1.) -> None:
-        super().__init__()
-
-        # Generic loss parameters.
-        self.reduction = reduction
-
-        # Loss-specific parameters.
-        if scale_weights is None:
-            scale_weights_from_ms_ssim_paper = [0.0448, 0.2856, 0.3001, 0.2363, 0.1333]
-            scale_weights = scale_weights_from_ms_ssim_paper
-        self.scale_weights = scale_weights if isinstance(scale_weights, torch.Tensor) else torch.tensor(scale_weights)
-        self.kernel_size = kernel_size
-        self.kernel_sigma = kernel_sigma
-        self.k1 = k1
-        self.k2 = k2
-        self.data_range = data_range
-
-    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
-        r"""Computation of Multi-scale Structural Similarity (MS-SSIM) index as a loss function.
-        The size of the image should be at least (kernel_size - 1) * 2 ** (levels - 1) + 1.
-        For colour images channel order is RGB.
-
-        Args:
-            prediction: Tensor with shape 2D (H, W), 3D (C, H, W), 4D (N, C, H, W) or 5D (N, C, H, W, 2).
-            target: Tensor with shape 2D (H, W), 3D (C, H, W), 4D (N, C, H, W) or 5D (N, C, H, W, 2).
-
-        Returns:
-            Value of MS-SSIM loss to be minimized, i.e. 1-`ms_sim`. 0 <= MS-SSIM loss <= 1. In case of 5D tensor,
-            complex value is returned as a tensor of size 2.
-        """
-
-        score = multi_scale_ssim(x=prediction, y=target, kernel_size=self.kernel_size, kernel_sigma=self.kernel_sigma,
-                                 data_range=self.data_range, reduction=self.reduction, scale_weights=self.scale_weights,
-                                 k1=self.k1, k2=self.k2)
-        return torch.ones_like(score) - score
-
-
 def _ssim_per_channel(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor,
                       data_range: Union[float, int] = 1., k1: float = 0.01,
                       k2: float = 0.03) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
```

Comment:
In case of avoiding Unions as much as possible, we should replace it with `List[torch.Tensor]`.